src/groupadd.c

```diff
@@ -120,6 +120,15 @@ usage (int status)
 	exit (status);
 }
 
+static void fail_exit(int status)
+{
+#ifdef WITH_AUDIT
+	audit_logger(AUDIT_ADD_GROUP, Prog, "add-group", group_name,
+				 AUDIT_NO_ID, SHADOW_AUDIT_FAILURE);
+#endif
+	exit (status);
+}
+
```

Comment:
Would it maybe be interesting to register the audit_logger() call with atexit(3)?  That would avoid changing so many lines from exit() to fail_exit().

Reply:
```
atexit(3)               Library Functions Manual              atexit(3)

NAME
     atexit - register a function to be called at normal process termi‐
     nation

LIBRARY
     Standard C library (libc, -lc)

SYNOPSIS
     #include <stdlib.h>

     int atexit(typeof(void (void)) *function);

DESCRIPTION
     The atexit() function registers the given function to be called at
     normal  process termination, either via exit(3) or via return from
     the program’s main().  Functions so registered are called  in  the
     reverse order of their registration; no arguments are passed.
```

Then, on successful exit, we could call `fflush(NULL); _exit(EXIT_SUCCESS);`, to not call the handler.

Reply:
```c
void
audit_atexit(void)
{
    audit_logger(AUDIT_ADD_GROUP, Prog, "add-group", group_name,
				 AUDIT_NO_ID, SHADOW_AUDIT_FAILURE);
}

atexit(&audit_atexit);
```

Reply:
The thing is that we don't always want to record an audit message when we exit this program, there are times when you just exit the program.